File: ModuleProjects/ParkingArea/UsingJAVA/src/main/java/com/kururu/simple/project/function/EndOfBusiness.java
package com.kururu.simple.project.function;

import static com.kururu.simple.project.constant.ParkingAreaEnums.END_BUSINESS_FLG;

import com.google.common.collect.Lists;
import com.kururu.simple.project.entity.EntryBook;
import com.kururu.simple.project.repository.EntryBookRepository;
import com.kururu.simple.project.repository.condition.BusinessStatusCondition;
import com.kururu.simple.project.utility.common.DateComponent;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.CollectionUtils;

import java.util.List;

/**
 * <h2>Parking Area [End of Business]</h2>
 *
 * @author Rick00Kim dreamx119@gmail.com
 */
@Component
@Slf4j
public class EndOfBusiness extends AbstractFunction {

    private static final String TARGET_ENTRY_BOOK_LIST = "targetEntryBookList";

    @Autowired
    private EntryBookRepository entryBookRepository;

    @Autowired
    private DateComponent dateComponent;

    @Override
    protected RESULT_STATUS input() {
        return RESULT_STATUS.SUCCESS;
    }

    @Override
    protected RESULT_STATUS validate() {
        /* Get target list of validate */
        final List<EntryBook> entryBookList
                = entryBookRepository.selectEntryBookAboutBusiness(BusinessStatusCondition.builder()
                .lotNumber(currentLotInformation.getLotNumber())
                .pairTermADay(dateComponent.getDateTermADay())
                .endBusinessFlgList(Lists.newArrayList(END_BUSINESS_FLG.values()))
                .build()
        );
        /* Validate */
        if (!CollectionUtils.isEmpty(entryBookList)) {
            entryBookList.forEach(e -> log.warn(String.format("Vehicle Number[%s] is not ended. please check", e.getKey().getVehicleNumber())));
            return RESULT_STATUS.FAILURE;
        }

        functionMap.put(TARGET_ENTRY_BOOK_LIST, entryBookList);

        return RESULT_STATUS.SUCCESS;
    }

    @Override
    public RESULT_STATUS process() {
        return RESULT_STATUS.SUCCESS;
    }

}

File: ModuleProjects/ParkingArea/UsingPython/parking_area/functions/end_of_business.py
from parking_area.functions.base_function import BaseFunction
from parking_area.database.entry_book_repo import EntryBookRepository
from parking_area.constants.parking_area_enums import ParkingAreaEnums
from parking_area.constants.parking_area_constants import ParkingAreaConstants
from parking_area.utilities.time_utility import get_start_on_today, get_end_on_today
from parking_area.utilities.log_utility import output_warn_log


class EndOfBusiness(BaseFunction):
    """Parking Area [End of Business]

    """

    function_name = "End of Business"

    def input_func(self):
        return ParkingAreaEnums.ResultStatusEnums.SUCCESS

    def validate_func(self):
        output_list = EntryBookRepository().select_for_income_file(
            (ParkingAreaConstants.CURRENT_LOT_INFORMATION.lot_number,
             get_start_on_today(),
             get_end_on_today(),
             '0')
        )

        if output_list:
            for warn_target in output_list:
                output_warn_log("Vehicle Number[%s] is not ended. please check" % warn_target.vehicle_number)
            return ParkingAreaEnums.ResultStatusEnums.FAILURE

        return ParkingAreaEnums.ResultStatusEnums.SUCCESS

    def process_func(self):
        return ParkingAreaEnums.ResultStatusEnums.SUCCESS
